Service/Shipment/Packingslip/Factory.php

```diff
@@ -101,29 +110,37 @@ public function __construct(
      */
     public function create($magentoShipment, $forceMagento = false)
     {
-        if (!$this->moduleManager->isEnabled('Fooman_PrintOrderPdf') || $forceMagento) {
+        if (!$this->moduleManager->isEnabled('Fooman_PdfCustomiser') || $forceMagento) {
             $renderer = $this->magentoPdf->getPdf([$magentoShipment]);
             // @codingStandardsIgnoreLine
             $this->setY($this->magentoPdf->y);
             return $renderer->render();
         }
 
-        return $this->getFoomanPdf($magentoShipment->getOrderId());
+        return $this->getFoomanPdf($magentoShipment);
     }
 
     /**
-     * @param $orderId
+     * @param ShipmentInterface $magentoShipment
      *
      * @return string
      * @throws NotFoundException
      */
-    private function getFoomanPdf($orderId)
+    private function getFoomanPdf($magentoShipment)
     {
-        $order = $this->orderRepository->get($orderId);
-        /** @var \Fooman\PdfCustomiser\Block\OrderShipmentFactory $documentManager */
-        // @codingStandardsIgnoreLine
-        $documentManager = $this->objectManager->create('\Fooman\PdfCustomiser\Block\OrderShipmentFactory');
-        $document = $documentManager->create(['data' => ['order' => $order]]);
+        if ($this->scopeConfig->isSetFlag('sales_pdf/shipment/shipmentuseorder')) {
+            $orderId = $magentoShipment->getOrderId();
+            $order = $this->orderRepository->get($orderId);
+            /** @var \Fooman\PdfCustomiser\Block\OrderShipmentFactory $documentManager */
+            // @codingStandardsIgnoreLine
+            $documentManager = $this->objectManager->create('\Fooman\PdfCustomiser\Block\OrderShipmentFactory');
```

Comment:
Is there any reason to use an object manager?

Reply:
@mcspronko thanks for checking this. The issue with ObjectManager is that all classes that are mentioned in the constructor are required during the `setup:di:compile` step. This becomes a problem for optional workarounds like this. In a lot of cases PdfCustomiser would not be installed so it would fail as the class is not found, hence having the class only requested at runtime when it's needed.

Unfortunately Composer can't help us here either as it does not offer a way to handle conditional dependencies. Ideally we would be able to instruct something like
if A is installed no changes 
if B installed no changes 
if A and B installed also install this extra compatibility module C

Reply:
@fooman thanks for your additional information. This must be the reason why it is implemented the way it is. I feel it should be documented in the code, though, to prevent further confusion. 

When I take up your PR, I'll take this up too.

Thanks for your input @mcspronko !

Reply:
A different option would be to create a separate extension that people should require if they use both the extensions. 

The question is how desirable this is, because people that don't read up documentation properly will still install both extensions and encounter issues, thus getting in touch with support.

Reply:
I think the getFoomanPdf should be moved to a minimum as a separate class. Today we have Fooman PDF, tomorrow there is going to be Pronko PDF, so the logic, implemented in the Factory class not quite flexible. It can be a better idea to implement MagentoFactory and FoomanFactory classes responsible for its own PDF provider.

Also, as a workaround, you guys can use Proxy class in a __construct() method as dependencies which are optional. So the actual instance of a class (Fooman\PdfCustomiser\Block\ShipmentFactory or Fooman\PdfCustomiser\Block\OrderShipmentFactory) will be created only when its actually used.

Reply:
@mcspronko I really like the materialised proxy suggestion. I believe this could become a pattern for these type of situations. I'll update this pull request with it.

@tig-dennisvanderhammen indeed having the end customer run into an issue first before installing the extra addon is the tricky part in all this. Might be a good feature for an ExtDN installer.

Reply:
@mcspronko @fooman Hmm, this is actually against another coding standard, which says not to hard-code any `Proxy` but use it only via `di.xml`. https://devdocs.magento.com/guides/v2.3/extension-dev-guide/proxies.html

Reply:
Yeap but maybe our discussion here leads to a PR to the above page?

There is currently no good way to deal with what is essentially a conditional dependency. The options we have collected so far would be:

1. use ObjectManager to retrieve classes at runtime
2. similar to above except do this in a Proxy class
3. create an additional module that customer needs to manually install 
4. in this particular case I could potentially create a plugin on the Factory class inside PdfCustomiser (as no classes from PostNL need to be instantiated) but the migration path on this is not clear to me and I am keen to figure out a generic way for these type of issues.

Any other options/ideas?

Reply:
In essence, a `Proxy` (and a `Factory`) is just a workaround for using the ObjectManager. So in the end, it boils down to the same thing. You could also create your own factory to produce the factories if they are there - the Factory-of-Factories pattern - and maybe involving the ObjectManager with a simple `class_exists`.

Reply:
Up until I came across the above mentioned coding standard guide (which I still like to understand the intention of) I liked the idea of Proxy as the stand-in nature seems fitting to the issue at hand.

Reply:
I have found a couple usages of materialised proxies in Magento core `Magento\Config\Model\Config\Structure\Element\Group\Proxy` and `Magento\Config\Model\Config\Structure\Search\Proxy`.

Reply:
Yup. I understand. But running unit tests (without the presence of Magento) would break dependencies with those materialised proxies, because they actually don't exist. Because of this, I've created hard-coded proxies myself, but I'm not really proud of those. Instead, a Factory-of-Factories class could be hard-coded to do the same work.